node/narwhal/src/gateway.rs

```diff
@@ -834,65 +834,132 @@ impl<N: Network> Gateway<N> {
+            f.debug_tuple("Gateway").field(&self.account.address()).field(&self.tcp.config()).finish()
         }
+    }
+
+    impl Arbitrary for Gateway<CurrentNetwork> {
+        type Parameters = ();
+        type Strategy = BoxedStrategy<Gateway<CurrentNetwork>>;
 
-        pub fn is_valid(&self) -> bool {
-            self.committee_input.is_valid() && self.workers_count < MAX_WORKERS
+        fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
+            any_valid_dev_gateway()
+                .prop_map(|(storage, _, account, dev)| Gateway::new(account, storage, dev).unwrap())
+                .boxed()
         }
+    }
+
+    type GatewayInput = (Storage<CurrentNetwork>, CommitteeContext, Account<CurrentNetwork>, Option<u16>);
+
+    fn any_valid_dev_gateway() -> BoxedStrategy<GatewayInput> {
+        (any::<CommitteeContext>(), any::<Selector>())
+            .prop_flat_map(|(context, account_selector)| {
+                let CommitteeContext(_, ValidatorSet(validators)) = context.clone();
+                (
+                    any_with::<Storage<CurrentNetwork>>(context.clone()),
+                    Just(context),
+                    Just(account_selector.select(validators)),
+                    0u8..,
+                )
+                    .prop_map(|(a, b, c, d)| (a, b, c.account, Some(d as u16)))
+            })
+            .boxed()
+    }
+
+    fn any_valid_prod_gateway() -> BoxedStrategy<GatewayInput> {
+        (any::<CommitteeContext>(), any::<Selector>())
+            .prop_flat_map(|(context, account_selector)| {
+                let CommitteeContext(_, ValidatorSet(validators)) = context.clone();
+                (
+                    any_with::<Storage<CurrentNetwork>>(context.clone()),
+                    Just(context),
+                    Just(account_selector.select(validators)),
+                )
+                    .prop_map(|(a, b, c)| (a, b, c.account, None))
+            })
+            .boxed()
+    }
 
-        pub async fn generate_workers(
-            &self,
-            gateway: &Gateway<CurrentNetwork>,
-        ) -> (IndexMap<u8, Worker<CurrentNetwork>>, IndexMap<u8, WorkerSender<CurrentNetwork>>) {
+    #[proptest]
+    fn gateway_dev_initialization(#[strategy(any_valid_dev_gateway())] input: GatewayInput) {
+        let (storage, _, account, dev) = input;
+        let address = account.address();
+        let gateway = Gateway::new(account, storage, dev);
+        assert!(gateway.is_ok());
+        let gateway = gateway.unwrap();
+        let tcp_config = gateway.tcp().config();
+        assert_eq!(tcp_config.listener_ip, Some(IpAddr::V4(Ipv4Addr::LOCALHOST)));
+        assert_eq!(tcp_config.desired_listening_port, Some(MEMORY_POOL_PORT + dev.unwrap()));
+
+        let tcp_config = gateway.tcp().config();
+        assert_eq!(tcp_config.max_connections, MAX_COMMITTEE_SIZE);
+        assert_eq!(gateway.account().address(), address);
```

Comment:
I think there's some unnecessary variable shadowing in this case? I'd perhaps write it like this? 
```suggestion
        let gateway = Gateway::new(account, storage, dev);
        let gateway = gateway.unwrap();
        assert_eq!(gateway.account().address(), address);
        
        let tcp_config = gateway.tcp().config();
        assert_eq!(tcp_config.listener_ip, Some(IpAddr::V4(Ipv4Addr::LOCALHOST)));
        assert_eq!(tcp_config.desired_listening_port, Some(MEMORY_POOL_PORT + dev.unwrap()));
        assert_eq!(tcp_config.max_connections, MAX_COMMITTEE_SIZE);
```